### PR TITLE
[SPARK-48767][SQL] Fix some error prompts when `variant` type data is invalid

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to:
- fix some error prompts when `variant` type data is invalid.
- provide a clear `error-condition` for `variant` related errors.
- use `checkError` to check exception in the `VariantSuite` class.

### Why are the changes needed?
- Reproduction examples
  image

  For reproduction examples detail, please refer to: https://github.com/apache/spark/pull/47162#issuecomment-2202255944

- When there is only `value` or `metadata` in `variant` data, it will throw `variant with more than two field`, which is obviously incorrect.
https://github.com/apache/spark/blob/930422389352b8349e5a845c8cae9993d30dce17/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala#L405-L408

- When `SQLConf.PARQUET_VECTORIZED_READER_ENABLED` is true or false, there is a difference in the error prompt, and we should align it.

### Does this PR introduce _any_ user-facing change?
Yes, fix some error prompts when `variant` type data is invalid.

### How was this patch tested?
Existed UT & Update UT.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
```
test("SPARK-47546: invalid variant binary") {
    // Write a struct-of-binary that looks like a Variant, but with minor variations that may make
    // it invalid to read.
    // Test cases:
    // 1) A binary that is almost correct, but contains an extra field "paths"
    // 2,3) A binary with incorrect field names
    // 4) Incorrect data typea
    // 5,6) Nullable value or metdata

    // Binary value of empty metadata
    val m = "X'010000'"
    // Binary value of a literal "false"
    val v = "X'8'"
    val cases = Seq(
      (s"named_struct('value', $v)",
        "INVALID_VARIANT.WRONG_NUM_FIELDS",
        "variant column must contain exactly two fields",
        "variant with more than two fields"),
      (s"named_struct('value', $v, 'metadata', $m, 'paths', $v)",
        "INVALID_VARIANT.WRONG_NUM_FIELDS",
        "variant column must contain exactly two fields",
        "variant with more than two fields"),
      (s"named_struct('value', $v, 'dictionary', $m)",
        "_LEGACY_ERROR_TEMP_1173",
        "variant missing metadata field",
        "variant missing metadata field"),
      (s"named_struct('val', $v, 'metadata', $m)",
        "_LEGACY_ERROR_TEMP_1173",
        "variant missing value field",
        "variant missing value field"),
      (s"named_struct('value', 8, 'metadata', $m)",
        "_LEGACY_ERROR_TEMP_1173",
        "variant column must be a non-nullable binary",
        "variant value must be a non-nullable binary"),
      (s"named_struct('value', cast(null as binary), 'metadata', $m)",
        "_LEGACY_ERROR_TEMP_1173",
        "variant column must be a non-nullable binary",
        "variant value must be a non-nullable binary"),
      (s"named_struct('value', $v, 'metadata', cast(null as binary))",
        "_LEGACY_ERROR_TEMP_1173",
        "variant column must be a non-nullable binary",
        "variant metadata must be a non-nullable binary")
    )
    cases.foreach { case (structDef, errorClass, msg1, msg2) =>
      Seq(false, true).foreach { vectorizedReader =>
        withSQLConf(SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key -> vectorizedReader.toString) {
          withTempDir { dir =>
            val file = new File(dir, "dir").getCanonicalPath
            val df = spark.sql(s"select $structDef as v from range(10)")
            df.write.parquet(file)
            val schema = StructType(Seq(StructField("v", VariantType)))
            val result = spark.read.schema(schema).parquet(file).selectExpr("to_json(v)")
            println(s"$structDef, $vectorizedReader")
            val e = intercept[org.apache.spark.SparkException](result.collect())
            checkError(
              exception = e.getCause.asInstanceOf[AnalysisException],
              errorClass = errorClass,
              parameters = Map("parquetType" -> (if (!vectorizedReader) msg1 else msg2))
            )
          }
        }
      }
    }
  }
```

**Comment by panbingkun** [comment]
cc @cashmand @cloud-fan

**Comment by cloud-fan** [comment]
cc @gene-db

**Comment by cashmand** [comment]
The changes look good to me.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 1]
```diff
@@ -2921,6 +2921,29 @@
     ],
     "sqlState" : "42K09"
   },
+  "INVALID_VARIANT" : {
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-conditions.json:2924]: shall we be more specific as this only happens when reading variant from parquet? something like `INVALID_VARIANT_FROM_PARQUET`

Comment (panbingkun) [reply common/utils/src/main/resources/error/error-conditions.json]: Yes, currently only `parquet`, `json` and `binary` formats are supported. 
(PS: json - read & write, binary - read
We can be suffixed with 'FROM-PARQUET'. 
If we encounter `more general scenarios later`, we can refactor and reuse it.

Updated.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**